onnxruntime/core/mlas/lib/jblas_gemm.cpp

```diff
@@ -0,0 +1,525 @@
+                    if (NTile == tAVX512F::NTILE && _cd->AVX512F()) {
+                        size = std::max(
+                            JblasSQ4GemmCompF32WorkspaceSize<tAVX512F>(
+                                M, N, K, DataParams[i].A, DataParams[i].lda, kptr, DataParams[i].C,
+                                DataParams[i].ldc
+                            ),
+                            size
+                        );
+                    } else if (NTile == tAVX2::NTILE && _cd->AVX2()) {
+                        size = std::max(
+                            JblasSQ4GemmCompF32WorkspaceSize<tAVX2>(
+                                M, N, K, DataParams[i].A, DataParams[i].lda, kptr, DataParams[i].C,
+                                DataParams[i].ldc
+                            ),
+                            size
+                        );
+                    }
+                }
+                if (CType == uint32_t(gemm::CompType::COMP_INT8_US_INT32)) {
+                    if (NTile == tAMX_INT8_US::NTILE && _cd->AMX_INT8()) {
+                        size = std::max(
+                            JblasSQ4GemmCompInt8WorkspaceSize<tAMX_INT8_US>(
+                                M, N, K, DataParams[i].A, DataParams[i].lda, kptr, DataParams[i].C,
+                                DataParams[i].ldc
+                            ),
+                            size
+                        );
+                    } else if (NTile == tAVX512_VNNI::NTILE && _cd->AVX512_VNNI()) {
+                        size = std::max(
+                            JblasSQ4GemmCompInt8WorkspaceSize<tAVX512_VNNI>(
+                                M, N, K, DataParams[i].A, DataParams[i].lda, kptr, DataParams[i].C,
+                                DataParams[i].ldc
+                            ),
+                            size
+                        );
+                    } else if (NTile == tAVX_VNNI::NTILE && _cd->AVX_VNNI()) {
+                        size = std::max(
+                            JblasSQ4GemmCompInt8WorkspaceSize<tAVX_VNNI>(
+                                M, N, K, DataParams[i].A, DataParams[i].lda, kptr, DataParams[i].C,
+                                DataParams[i].ldc
+                            ),
+                            size
+                        );
+                    }
+                }
+                if (CType == uint32_t(gemm::CompType::COMP_INT8_SS_INT32)) {
+                    if (NTile == tAMX_INT8_SS::NTILE && _cd->AMX_INT8()) {
+                        size = std::max(
+                            JblasSQ4GemmCompInt8WorkspaceSize<tAMX_INT8_SS>(
+                                M, N, K, DataParams[i].A, DataParams[i].lda, kptr, DataParams[i].C,
+                                DataParams[i].ldc
+                            ),
+                            size
+                        );
+                    }
+                }
+            }
+            delete ptr;
```

Comment:
is all the code between the assignment to `ptr` and this `delete` guaranteed to never throw any uncaught exceptions?
even if so, it places some burden on people working on this code in the future to know about this.
would it be better to use some RAII type (e.g. `std::unique_ptr`) to handle deleting reliably?

Reply:
Yes, there is no exception between those codes. But I think your suggestion is better, I will use std::unique_ptr instead.

Reply:
changed